hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/burn/BurnCall.java

```diff
@@ -16,6 +16,69 @@
+import com.hedera.node.app.service.contract.impl.exec.gas.DispatchType;
+import com.hedera.node.app.service.contract.impl.exec.gas.SystemContractGasCalculator;
+import com.hedera.node.app.service.contract.impl.exec.scope.SystemContractOperations;
+import com.hedera.node.app.service.contract.impl.exec.scope.VerificationStrategy;
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.AddressIdConverter;
 import com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.HtsCall;
+import com.hedera.node.app.service.token.records.TokenBurnRecordBuilder;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.function.LongFunction;
+import java.util.function.Supplier;
+
+public interface BurnCall extends HtsCall {
+    /**
+     * Performs the {@link HtsCall#execute()} logic shared between fungible and non-fungible burns.
+     *
+     * @param tokenId the token ID, if found
+     * @param bodySupplier the function to create the transaction body
+     * @param dispatchType the dispatch type
+     * @param addressIdConverter the address ID converter
+     * @param verificationStrategy the verification strategy
+     * @param gasCalculator the gas calculator
+     * @param spender the spender
+     * @param systemContractOperations the system contract operations
+     * @param invalidTokenReversion the invalid token reversion
+     * @return the result of the call
+     */
 
-public interface BurnCall extends HtsCall {}
+    // too may parameters
+    @SuppressWarnings("java:S107")
+    default @NonNull PricedResult executeBurn(
+            @Nullable final TokenID tokenId,
+            @NonNull final DispatchType dispatchType,
+            @NonNull final AddressIdConverter addressIdConverter,
+            @NonNull final VerificationStrategy verificationStrategy,
+            @NonNull final SystemContractGasCalculator gasCalculator,
+            @NonNull final Supplier<TransactionBody> bodySupplier,
+            @NonNull final org.hyperledger.besu.datatypes.Address spender,
+            @NonNull final SystemContractOperations systemContractOperations,
+            @NonNull final LongFunction<PricedResult> invalidTokenReversion) {
+        if (tokenId == null) {
+            return invalidTokenReversion.apply(gasCalculator.canonicalPriceInTinybars(dispatchType));
+        }
+        final var spenderId = addressIdConverter.convert(asHeadlongAddress(spender.toArrayUnsafe()));
+        final var body = bodySupplier.get();
+        final var gasRequirement = gasCalculator.gasRequirement(body, dispatchType);
+        final var recordBuilder =
+                systemContractOperations.dispatch(body, verificationStrategy, spenderId, TokenBurnRecordBuilder.class);
+        if (recordBuilder.status() != ResponseCodeEnum.SUCCESS) {
+            return gasOnly(revertResult(recordBuilder.status(), gasRequirement));
+        } else {
```

Comment:
(Personally I'd leave off the `else` and just put the next 4 lines at the main level of the method.  To set off the error return from the normal path. But maybe that's just me.)

Reply:
Done.